File: easy/1351_count_negative_numbers_in_a_sorted_matrix/solution.cpp
using namespace std;

#include <vector>

class Solution {
public:
  int countNegatives(vector<vector<int>> &grid) {
    const int cols = grid[0].size();

    int left, right, mid;
    int negativesCount = 0;

    for (const vector<int> &row : grid) {
      left = 0;
      right = cols - 1;

      while (left <= right) {
        mid = left + (right - left) / 2;

        if (row[mid] < 0 && (mid == 0 or row[mid - 1] >= 0)) {
          negativesCount += cols - mid;
          break;
        } else if (row[mid] < 0) {
          right = mid - 1;
        } else {
          left = mid + 1;
        }
      }
    }

    return negativesCount;
  }
};

File: easy/1351_count_negative_numbers_in_a_sorted_matrix/solution.py
from typing import List


class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        cols = len(grid[0])
        rows = len(grid)

        negatives_count = 0

        for row in range(rows):
            for col in range(cols - 1, -1, -1):
                if grid[row][col] >= 0:
                    break

                negatives_count += 1

        return negatives_count
